src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg.py

```diff
@@ -1030,92 +1007,65 @@ def visit_Program(self, node: gtir.Program) -> dace.SDFG:
-            assert not target_desc.transient
-
             assert source.gt_type == target.gt_type
             field_domain = gtir_domain.get_field_domain(target_domain)
             source_subset = _make_access_index_for_field(field_domain, source)
             target_subset = _make_access_index_for_field(field_domain, target)
 
-            if target.dc_node.data in state_input_data:
-                # create new access nodes in the target state
-                target_state.add_nedge(
-                    target_state.add_access(source.dc_node.data),
-                    target_state.add_access(target.dc_node.data),
-                    dace.Memlet(
-                        data=target.dc_node.data, subset=target_subset, other_subset=source_subset
-                    ),
-                )
-                # remove isolated access node
-                state.remove_node(target.dc_node)
-            else:
-                state.add_nedge(
-                    source.dc_node,
-                    target.dc_node,
-                    dace.Memlet(
-                        data=target.dc_node.data, subset=target_subset, other_subset=source_subset
-                    ),
-                )
+            target_state.add_nedge(
+                # create in the target state new access nodes to the field operator result
+                target_state.add_access(source.dc_node.data),
+                target.dc_node,
+                dace.Memlet(
+                    data=target.dc_node.data, subset=target_subset, other_subset=source_subset
+                ),
+            )
+            if ctx.state.degree(source.dc_node) == 0:
+                ctx.state.remove_node(source.dc_node)
+
+        # Visit the domain expression.
+        domain = gtir_domain.extract_target_domain(stmt.domain)
+
+        # Visit the field operator expression.
+        source_tree = self._visit_expression(stmt.expr, ctx)
 
+        # In order to support inout argument, write the result in separate next state
+        # this is needed to avoid undefined behavior for expressions like: X, Y = X + 1, X
+        target_state = ctx.sdfg.add_state_after(ctx.state, f"post_{ctx.state.label}")
+        # The target expression could be a `SymRef` to an output field or a `make_tuple`
+        # expression in case the statement returns more than one field.
+        target_tree = self._visit_expression(
+            stmt.target,
+            ctx=SubgraphContext(ctx.sdfg, target_state, ctx.scope_symbols),
+            use_temp=False,
+        )
         gtx_utils.tree_map(
             lambda source, target, target_domain: _visit_target(
                 source, target, target_domain, target_state
             )
         )(source_tree, target_tree, domain)
 
-        if target_state.is_empty():
-            sdfg.remove_node(target_state)
-            return state
-        else:
-            return target_state
+        return target_state
```

Comment:
I think you should return a context here
```suggestion
        return SubgraphContext(sdfg, target_state)
```
I think this could be more consistent in the overall design.

Reply:
I tried to use it everywhere, but here it would make things more complicated than needed, I think.